### PR TITLE
[flexsn] Clean up HOP and compile training improvements

### PR DESCRIPTION
**Pull request by fangwei123456**
## Summary
- create a clean FlexSN/HOP release branch from current `master`
- keep only the files relevant to FlexSN HOP and compile-training work
- carry over the validated FlexSN Inductor/HOP tests and Conv-BN fusion regression test
- harden optional HOP imports and skip Windows-only fullgraph Inductor tests that fail for platform reasons

## Validation
- `C:\Users\fw\anaconda3\envs\torch\python.exe -m pytest test\activation_based\test_flex_sn_inductor.py test\activation_based\test_fuse_conv_bn.py -q`
- `py_compile` on the touched FlexSN benchmark / kernel / test files

## Notes
- This PR supersedes #662 with the same core FlexSN scope but without the unrelated history that had accumulated on `feature/flexsn-hop`.

## Summary by CodeRabbit

* **New Features**
  * Conv–BatchNorm eval/train transforms, HOP/scan helpers, Dynamo HOP integration, and example-driven output templates with explicit empty-sequence (T==0) fast paths.

* **Performance**
  * Per-iteration CUDA benchmarking with optional reset hooks for consistent timing; deterministic compiled-kernel caching and reuse.

* **Bug Fixes**
  * Correct handling of zero-length sequences, final-state/gradient edge cases, and stricter dtype/device/template validation.

* **Refactor**
  * Simplified kernel/build and unified inference/training/backward flows; template-driven wrapper/trace plumbing.

* **Tests**
  * Expanded coverage for fusion, T==0 semantics, multi-backend and compilation/Dynamo behaviors.

### CODE REVIEW

**Comment by fangwei123456** [comment]
@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed fixes for the current review round:
- benchmark timing now excludes reset work
- benchmark header label/f-string cleanup restored
- `flex_sn_inductor.__init__` now exports an explicit `__all__`
- `torch2graph` backward tracing now includes all differentiable outputs
- restored `flexsn_backward_ncl_bucket` compatibility shim for the bucket benchmark
- added the requested public API doc/type updates

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the remaining actionable review feedback in `1647167f`.

This round fixes the real correctness guards and clears the low-risk API/doc/style issues that were still generating review noise:
- restored Triton/Inductor scan validation so all `example_inputs` must match the first example tensor's element count (and dtype)
- added a fast-fail `requires_grad` length check in `torch2graph.generate_forward_and_backward_graph(...)`
- updated public typing/docstrings for the touched helpers and wrappers
- sorted `flex_sn_inductor.__all__`
- hoisted the benchmark-local import and fully typed both `cuda_time_ms(...)` helpers
- cleaned the new Sphinx field text to use ASCII punctuation and documented the second Conv-BN fusion `ValueError` path

Validation:
- `C:\Users\fw\anaconda3\envs\torch\python.exe -m pytest test\activation_based\test_flex_sn_inductor.py test\activation_based\test_fuse_conv_bn.py -q`
  - `103 passed, 15 skipped`
- `py_compile` passed for all touched benchmark/kernel/test files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the two remaining actionable review threads in `0fe2df80`.

This round:
- synchronized the public `example_outputs` docs in `FlexSN` with the actual backend-specific contract (`torch` requires it for `T == 0`; scan backends validate against the first example input's per-step shape/dtype)
- fixed `flexsn_backward(...)` so input-sequence gradients preserve the original input templates even when `num_outputs == 0` and state shapes are only numel-compatible
- added regression tests for both the normal and `T == 0` state-only backward cases

Validation:
- `C:\Users\fw\anaconda3\envs\torch\python.exe -m pytest test\activation_based\test_flex_sn_inductor.py test\activation_based\test_fuse_conv_bn.py -q`
  - `105 passed, 15 skipped`
- `py_compile` passed for the touched files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the latest FlexSN review findings in `369ba453`.

Changes:
- stripped runtime HOP output template specs down to shape/dtype so empty-sequence outputs follow the actual input device instead of the constructor-time `example_outputs` device;
- taught the eager Triton wrapper to save and reuse both input-sequence and initial-state templates during backward, including the all-`None` gradient path;
- threaded combined input/state templates through the inductor custom backward op so compiled training now preserves the same state-gradient shapes as the eager wrapper;
- added targeted regressions for HOP template devices, eager state-template gradients, all-`None` incoming grads, and compiled backward template forwarding.

Validation:
- `C:\Users\fw\anaconda3\envs\torch\python.exe -m pytest test\\activation_based\\test_flex_sn_inductor.py test\\activation_based\\test_fuse_conv_bn.py -q`
- result: `109 passed, 15 skipped`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the remaining FlexSN findings in `97352dab`.

Changes:
- scan-backend `example_outputs` validation now runs against the same synthesized `example_inputs` used by the contract checker, so omitted `example_inputs` can no longer bypass shape/dtype validation;
- updated both Chinese and English `example_outputs` docs to reflect the backend-specific contract: `torch` needs templates for `T == 0`, `triton/inductor` validate against the first example input, and `hop` intentionally allows arbitrary templates that are materialized on the runtime input device;
- completed the public `flexsn_backward(...)` signature/docstring for the new `input_templates` / `state_templates` behavior and the all-`None` gradient path;
- added a regression for the omitted-`example_inputs` validation case.

Validation:
- `C:\Users\fw\anaconda3\envs\torch\python.exe -m pytest test\\activation_based\\test_flex_sn_inductor.py test\\activation_based\\test_fuse_conv_bn.py -q`
- result: `110 passed, 15 skipped`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the latest actionable review points and pushed `17345c07`.

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py test/activation_based/test_fuse_conv_bn.py -q`
  - `111 passed, 15 skipped`
- `py_compile` for the touched FlexSN files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the new actionable review points and pushed `76032207`.

This round:
- zero-fills mixed `None` gradient slots before launching the Triton backward kernel
- adds a regression for mixed output/state gradients
- tightens the HOP monkeypatch delegation so bound `make()` implementations keep the subclass context

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py test/activation_based/test_fuse_conv_bn.py -q`
  - `112 passed, 15 skipped`
- `py_compile` for the touched HOP / wrapper / test files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Addressed the latest actionable FlexSN review points and pushed `df1bff23`.

This round:
- validates empty-sequence scan core arity on both eager fast paths
- tightens `flexsn_backward(...)` so missing output-sequence grads require explicit input templates
- syncs the public HOP helper docstrings with the `output_template_specs` contract

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py test/activation_based/test_fuse_conv_bn.py -q`
  - `115 passed, 15 skipped`
- `py_compile` for the touched HOP / wrapper / test files

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed one more small cleanup in `60d2cccd` to preserve the original exception context when `output_template_specs` is not a Python constant during HOP lowering.

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py -q`
  - `102 passed, 15 skipped`
- `py_compile` for `spikingjelly/activation_based/triton_kernel/flex_sn_inductor/hop.py`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed one more tiny cleanup in `1beea32e` to normalize the remaining FlexSN HOP type annotation style in `hop.py`.

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py -q`
  - `102 passed, 15 skipped`
- `py_compile` for `spikingjelly/activation_based/triton_kernel/flex_sn_inductor/hop.py`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed another small FlexSN HOP follow-up in `47ddc3f4`.

This round:
- stops trace-time `output_template_specs` from pinning empty-scan outputs to the traced device
- keeps empty outputs on the runtime input device when Dynamo inferred the templates
- adds bilingual docstrings for the public availability helpers in `hop.py`

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py -q`
  - `102 passed, 15 skipped`
- `py_compile` for `spikingjelly/activation_based/triton_kernel/flex_sn_inductor/hop.py` and `test/activation_based/test_flex_sn_inductor.py`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
Pushed another FlexSN HOP review cleanup in `6d607e29`.

This round:
- rejects empty-scan fast paths for cores with required keyword-only parameters
- gates the Dynamo registration test on the actual HOP availability helper
- normalizes the new public helper docstrings to use repository-standard `:return:` fields

Validated with:
- `pytest test/activation_based/test_flex_sn_inductor.py -q`
  - `102 passed, 15 skipped`
- `py_compile` for `spikingjelly/activation_based/triton_kernel/flex_sn_inductor/hop.py` and `test/activation_based/test_flex_sn_inductor.py`

@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
@coderabbitai review

@gemini-code-assist review

@codex review

**Comment by fangwei123456** [comment]
@coderabbitai review

@gemini-code-assist review

@codex review
